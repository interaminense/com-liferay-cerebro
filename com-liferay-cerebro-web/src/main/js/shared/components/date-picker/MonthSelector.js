import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';
import getCN from 'classnames';
import moment from 'moment';
import {noop, range} from 'lodash';

import Button from 'shared/components/Button';
import Icon from 'shared/components/Icon';

class Month extends Component {
	@autobind
	handleClick() {
		const {onClick, value} = this.props;

		onClick(value);
	}

	render() {
		const {active, label} = this.props;

		return (
			<div class="month-root">
				<Button
					display="unstyled"
					elementClasses={getCN({active})}
					onClick={this.handleClick}
				>
					{label}
				</Button>
			</div>
		);
	}
}

Month.PROPS = {
	active: Config.bool().value(false),
	label: Config.string().required(),
	onClick: Config.func().required(),
	value: Config.number().required()
};

class Year extends Component {
	@autobind
	handleClick(month) {
		const {onSelect, value} = this.props;

		onSelect({month, year: value});
	}

	@autobind
	handleToggleExpand() {
		this.state.expanded = !this.state.expanded;
	}

	render() {
		const {props: {currentMonth, value}, state: {expanded}} = this;

		const classes = getCN('year-root', {expanded});

		return (
			<div class={classes}>
				<div class="year-header" onClick={this.handleToggleExpand}>
					{value}

					<Icon symbol={expanded ? 'caret-bottom' : 'caret-top'} />
				</div>

				{expanded && (
					<div class="year-body">
						{moment
							.monthsShort()
							.map((month, i) => (
								<Month
									active={
										currentMonth.year() === value &&
										currentMonth.month() === i
									}
									key={i}
									label={month}
									onClick={this.handleClick}
									value={i}
								/>
							))}
					</div>
				)}
			</div>
		);
	}
}

Year.STATE = {
	expanded: Config.bool().value(true)
};

Year.PROPS = {
	currentMonth: Config.validator(moment.isMoment).required(),
	onSelect: Config.func().value(noop),
	value: Config.number().required()
};

class MonthSelector extends Component {
	render() {
		const {minDate, currentMonth, onSelect} = this.props;

		const currentYear = moment().year();

		const startYear = minDate.year();

		return (
			<div class="month-selector-root">
				{range(
					currentYear,
					Math.min(startYear, currentYear - 1),
					-1
				).map(year => (
					<Year
						currentMonth={currentMonth}
						key={year}
						onSelect={onSelect}
						value={year}
					/>
				))}
			</div>
		);
	}
}

MonthSelector.PROPS = {
	currentMonth: Config.validator(moment.isMoment).required(),
	minDate: Config.validator(moment.isMoment).valueFn(moment),
	onSelect: Config.func().value(noop)
};

export default MonthSelector;