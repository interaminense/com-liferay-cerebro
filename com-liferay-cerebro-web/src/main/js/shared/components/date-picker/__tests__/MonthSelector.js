import dom from 'metal-dom';
import moment from 'moment';

import MonthSelector from '../MonthSelector';

describe('MonthSelector', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render', () => {
		component = new MonthSelector({
			currentMonth: moment(0).startOf('month')
		});

		expect(component).toMatchSnapshot();
	});

	it('should call onSelect when a month is selected', () => {
		const onSelect = jest.fn();

		component = new MonthSelector({
			currentMonth: moment(0).startOf('month'),
			onSelect
		});

		dom.triggerEvent(
			component.element.querySelector('.month-root button'),
			'click'
		);

		expect(onSelect).toHaveBeenCalled();
	});
});