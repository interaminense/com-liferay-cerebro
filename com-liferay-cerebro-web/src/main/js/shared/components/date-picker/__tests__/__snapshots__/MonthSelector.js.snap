// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthSelector should render 1`] = `
<div class="month-selector-root">
  <div class="year-root expanded">
    <div class="year-header">2017
      <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
        <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
      </svg>
    </div>
    <div class="year-body">
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">Jan</button>
      </div>
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">Feb</button>
      </div>
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">Mar</button>
      </div>
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">Apr</button>
      </div>
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">May</button>
      </div>
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">Jun</button>
      </div>
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">Jul</button>
      </div>
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">Aug</button>
      </div>
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">Sep</button>
      </div>
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">Oct</button>
      </div>
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">Nov</button>
      </div>
      <div class="month-root">
        <button class="button-root btn btn-unstyled" type="button">Dec</button>
      </div>
    </div>
  </div>
</div>
`;
